File: src/Java/pipeline-core/src/main/java/com/pipeline/core/PipelineExecution.java
package com.pipeline.core;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Objects;

/** Execution-scoped control available to Actions declared anywhere. */
public final class PipelineExecution {
  private static final ThreadLocal<Deque<ExecutionState<?>>> ACTIVE_EXECUTIONS =
      ThreadLocal.withInitial(ArrayDeque::new);

  private PipelineExecution() {}

  /** Marks the innermost actively executing Pipeline Action as short-circuited. */
  public static void shortCircuit() {
    ExecutionState<?> executionState = currentState();
    if (!executionState.isActionExecuting()) {
      throw new IllegalStateException(
          "shortCircuit() can only be called while a Pipeline Action is executing");
    }
    executionState.requestShortCircuit();
  }

  static Scope open(ExecutionState<?> executionState) {
    Objects.requireNonNull(executionState, "executionState");
    Deque<ExecutionState<?>> executions = ACTIVE_EXECUTIONS.get();
    executions.push(executionState);
    return new Scope(executionState);
  }

  static ExecutionState<?> currentState() {
    Deque<ExecutionState<?>> executions = ACTIVE_EXECUTIONS.get();
    if (executions.isEmpty()) {
      ACTIVE_EXECUTIONS.remove();
      throw new IllegalStateException(
          "shortCircuit() can only be called during an active pipeline run");
    }
    return executions.peek();
  }

  @SuppressWarnings("unchecked")
  static <C> ExecutionState<C> currentStateTyped() {
    return (ExecutionState<C>) currentState();
  }

  static final class Scope implements AutoCloseable {
    private final ExecutionState<?> expectedState;
    private boolean closed;

    private Scope(ExecutionState<?> expectedState) {
      this.expectedState = expectedState;
    }

    @Override
    public void close() {
      if (closed) return;
      closed = true;

      Deque<ExecutionState<?>> executions = ACTIVE_EXECUTIONS.get();
      if (executions.isEmpty() || executions.peek() != expectedState) {
        ACTIVE_EXECUTIONS.remove();
        throw new IllegalStateException("Pipeline execution scopes were closed out of order");
      }

      executions.pop();
      if (executions.isEmpty()) ACTIVE_EXECUTIONS.remove();
    }
  }
}

File: src/CSharp/pipeline_services/core/PipelineExecution.cs
using System;
using System.Threading;

namespace PipelineServices.Core;

internal interface IPipelineExecutionState
{
    bool ActionExecuting { get; }
    int ActionThreadId { get; }
    void RequestShortCircuit();
}

/// <summary>Execution-scoped control available to Actions declared anywhere.</summary>
public static class PipelineExecution
{
    private sealed class ExecutionFrame
    {
        internal ExecutionFrame(IPipelineExecutionState state, ExecutionFrame? parent)
        {
            State = state;
            Parent = parent;
        }

        internal IPipelineExecutionState State { get; }
        internal ExecutionFrame? Parent { get; }
    }

    private static readonly AsyncLocal<ExecutionFrame?> CurrentFrame = new();

    public static void ShortCircuit()
    {
        ExecutionFrame frame = CurrentFrame.Value
            ?? throw new InvalidOperationException(
                "ShortCircuit() can only be called during an active Pipeline run");

        IPipelineExecutionState state = frame.State;
        if (!state.ActionExecuting || state.ActionThreadId != Environment.CurrentManagedThreadId)
        {
            throw new InvalidOperationException(
                "ShortCircuit() can only be called while the current Pipeline Action body is executing");
        }
        state.RequestShortCircuit();
    }

    internal static IDisposable Open(IPipelineExecutionState state)
    {
        ArgumentNullException.ThrowIfNull(state);
        ExecutionFrame frame = new(state, CurrentFrame.Value);
        CurrentFrame.Value = frame;
        return new ExecutionScope(frame);
    }

    private sealed class ExecutionScope : IDisposable
    {
        private readonly ExecutionFrame expectedFrame;
        private bool disposed;

        internal ExecutionScope(ExecutionFrame expectedFrame)
        {
            this.expectedFrame = expectedFrame;
        }

        public void Dispose()
        {
            if (disposed)
            {
                return;
            }
            disposed = true;

            if (!ReferenceEquals(CurrentFrame.Value, expectedFrame))
            {
                CurrentFrame.Value = null;
                throw new InvalidOperationException(
                    "Pipeline execution scopes were closed out of order");
            }
            CurrentFrame.Value = expectedFrame.Parent;
        }
    }
}
